src/instance.rs

```diff
@@ -32,21 +32,142 @@ pub unsafe trait PyNativeType: Sized {
     }
 }
 
-/// A Python object of known type T.
+/// A GIL-independent reference to an object allocated on the Python heap.
 ///
-/// Accessing this object is thread-safe, since any access to its API requires a `Python<'py>` GIL
-/// token. There are a few different ways to use the Python object contained:
-///  - [`Py::as_ref`](#method.as_ref) to borrow a GIL-bound reference to the contained object.
-///  - [`Py::borrow`](#method.borrow), [`Py::try_borrow`](#method.try_borrow),
-///    [`Py::borrow_mut`](#method.borrow_mut), or [`Py::try_borrow_mut`](#method.try_borrow_mut),
-///    to directly access a `#[pyclass]` value (which has RefCell-like behavior, see
-///    [the `PyCell` guide entry](https://pyo3.rs/main/class.html#pycell-and-interior-mutability)
-///    ).
-///  - Use methods directly on `Py`, such as [`Py::call`](#method.call) and
-///    [`Py::call_method`](#method.call_method).
+/// This type does not auto-dereference to the inner object because you must prove you hold the GIL to access it.
+/// Instead, call one of its methods to access the inner object:
+///  - [`Py::as_ref`], to borrow a GIL-bound reference to the contained object.
+///  - [`Py::borrow`], [`Py::try_borrow`], [`Py::borrow_mut`], or [`Py::try_borrow_mut`],
+/// to get a (mutable) reference to a contained pyclass, using a scheme similar to std's [`RefCell`].
+/// See the [`PyCell` guide entry](https://pyo3.rs/main/class.html#pycell-and-interior-mutability)
+/// for more information.
+///  - You can call methods directly on `Py` with [`Py::call`], [`Py::call_method`] and friends.
+/// These require passing in the [`Python<'py>`](crate::Python) token but are otherwise similar to the corresponding
+/// methods on [`PyAny`].
 ///
-/// See [the guide](https://pyo3.rs/main/types.html) for an explanation
-/// of the different Python object types.
+/// # Example: Storing Python objects in structs
+///
+/// As all the native Python objects only appear as references, storing them in structs doesn't work well.
+/// For example, there is no way to make something like this compile!
+///
+/// ```compile_fail
+/// # use pyo3::prelude::*;
+/// # use pyo3::types::PyDict;
+/// #
+/// struct Foo<'py> {
+///     inner: &'py PyDict,
+/// }
+///
+/// impl Foo {
+///     fn new() -> Foo {
+///         let foo = Python::with_gil(|py| {
+///             // `py` will only last for this scope.
+///
+///             // `&PyDict` derives its lifetime from `py` and
+///             // so won't be able to outlive this closure.           
+///             let dict: &PyDict = PyDict::new(py);
+///
+///             // because `Foo` contains `dict` its lifetime
+///             // is now also tied to `py`.
+///             Foo { inner: dict }
+///         });
+///         // Foo is no longer valid.
+///         // Returning it from this function is a 💥 compiler error 💥
+///         foo
```

Comment:
Might want to use a `#[pyclass]`, because otherwise this could be made to compile by adding `py: Python` argument to `Foo::new`.

(I've actually used this exact pattern to bind temporarily to Python-owned data just for a short moment, it's also essentially what `#[derive(FromPyObject)]` does if you put a `&PyDict` reference in it.)

Reply:
I think it would be enough to show
```
/// #[pyclass]
/// struct Foo<'py> {
///     inner: &'py PyDict,
/// }
```
without the impl block.

Structs used purely in Rust can of course have a `'py` lifetime attached and contain `&'py` members, they just can't "escape" to Python.

Reply:
> Might want to use a #[pyclass]

Done.

> without the impl block.

I think the impl block is important to show that the lifetime will always be bound to the Python token used to create it. I would rather *show* this than explain it.

Also, while it is "lifetimes 101" and should be obvious to most Rust programmers, I think it's important to explain this in detail because I think a lot of people that use pyo3 are not experienced Rust programmers.

Reply:
Ok, fair enough.